### DISCUSSION TITLE
Virtual Environments & the '-m' Flag

### DISCUSSION THREAD

**Discussion by rontarrant** [beeware/toga#2679] [General]
Hi all,
Just discovered BeeWare and I think I'm in love. Greetings and salutations to the community.

I do, however, have a couple of questions arising from the first few pages of the Toga Tutorial in the docs.

1) I see this a lot where it's recommended that Python libraries are installed and used in virtual environments. Why is this so highly recommended? I get the whole compartmentalization thing, but do things really get out of hand if we don't use venv? And if so, how do they get out of hand? Or is this just an MS Windows thing?

2) The docs also say that BeeWare scripts should be run thusly:
`python -m helloworld`
But I tried it both ways (`python hellowworld.py`) and it seems the same to me. Or are the differences only apparent in larger, more complex, scripts?

**Comment by freakboy3742** [comment]
> Hi all, Just discovered BeeWare and I think I'm in love. Greetings and salutations to the community.

Glad you're enjoying your BeeWare experience :-)
 
> 1. I see this a lot where it's recommended that Python libraries are installed and used in virtual environments. Why is this so highly recommended? I get the whole compartmentalization thing, but do things really get out of hand if we don't use venv? And if so, how do they get out of hand? Or is this just an MS Windows thing?

It is entirely a compartmentalisation thing, and isn't Windows-specific at all. 

Do things get out of hand? Yes. Very quickly, and very easily.

Imagine you're working on two projects. One project requires setuptools < 58; one requires setuptools > 66. You literally can't resolve this in a single "monolithic" Python environment.

Imagine the same thing, but the dependency causing a conflict isn't a library you're installing directly, but 5 dependencies deep in a chain. Same problem... now try to diagnose and fix it.

Or, even if you're only working on one project: You're not quite sure how you're going to implement some new feature, so you install a library off PyPI. Which installs 3 other libraries. Which install 3 other libraries. Then you realise the original library doesn't do what you want. Now get yourself back to a clean state. 

In both cases, having a virtual environment is invaluable. Need 2 different versions of a library? Use 2 different virtual environments. Need to walk back an install? Delete the virtual environment and re-create it. Neither of these options are easy if you've got a single monolithic Python install.

Is it *possible* to use Python without virtual environments? Sure. It's also *possible* to cut your hair while standing up in a hammock. But there are much easier (and safer) ways to do it :-)

> 2. The docs also say that BeeWare scripts should be run thusly:
>    `python -m helloworld`
>    But I tried it both ways (`python hellowworld.py`) and it seems the same to me. Or are the differences only apparent in larger, more complex, scripts?

Yes - it's mostly for the benefit of more complex scripts.

When all you've got is a single file, `python -m helloworld` and `python helloworld.py` are essentially identical.

However, `python -m helloworld` will *also* work when `helloworld` is a *directory*. As long as that directory contains `__init__.py` (indicating it's a python module) and `__main__.py` (providing the entry point for the module), `python -m helloworld` will run that folder-as-a-module. As soon as you've got more code than fits in a single file, this gets really useful; and it's a lot more predictable at runtime because you don't need to configure your PYTHONPATH so that `python mymodule/app.py` can find `mymodule/otherfile.py` as well - if it could find `-m mymodule`, then it can find everything it needs.

**Reply by rontarrant** [reply]
> It's also possible to cut your hair while standing up in a hammock. But there are much easier (and safer) ways to do it :-)

Great analogy! :)
Thanks for all the info.
I went back to a monolithic Python environment while working on a series of demos for kivy. It seemed like a waste of drive space (not to mention the extra time/trouble) to set up environments for them all. But, maybe it's time to go back to a more sane set-up before I end up down some dark road with no paddle to steer by (to mix a few metaphors).
